chain-ser/src/packer.rs

```diff
@@ -64,79 +65,89 @@ impl<R: std::io::BufRead> Codec<R> {
     #[inline]
     pub fn get_bytes(&mut self, n: usize) -> Result<Vec<u8>, ReadError> {
         let mut buf = vec![0u8; n];
-        self.0.read_exact(&mut buf)?;
+        self.inner.read_exact(&mut buf)?;
         Ok(buf)
     }
     #[inline]
-    pub fn get_slice(&mut self, slice: &mut [u8]) -> Result<(), ReadError> {
-        self.0.read_exact(slice)?;
+    /// This is a wrapper over the std::io::BufRead::fill_buf() function,
+    /// so be aware of that you need also execute consume() function to move the reader position
+    pub fn get_slice(&mut self, n: usize) -> Result<&[u8], ReadError> {
+        let data = self.inner.fill_buf()?;
+        if data.len() < n {
+            return Err(ReadError::NotEnoughBytes(data.len(), n));
+        }
```

Comment:
We bump into another problem here. Having less than `n` bytes in a `BufRead` doesn't really _mean_ that there are no `n` bytes and we will never have these `n` bytes. Consider reading from a network stream wrapped into `BufReader` for example. And I do not see how we can bypass this issue without introducing some suboptimal approach. For example, we could call `BufRead::fill_buf()` until we get the required amount of bytes, but if the network stream is stalled we end up iterating forever. This is pretty much equivalent to hanging `read_exact` except the latter may trigger some timeout error in such case.

I am really starting to lean towards not removing the old implementation of deserialization from in-memory slices precisely because it accounts for all the specifics of in-memory buffers while not having to deal with the specifics of readers.